### PR TITLE
Make handler function for `/config` endpoint overrideable

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does**:

In order to be able to extend or replace the default handler function for the
`/config` endpoint, the registered handler evaluates whether a handler override
is set. If so, the custom handler function is invoked, otherwise the default
handler is used.

An example for this use-case is, that a downstream project extends the Cortex
configuration with additional configuration options, which should also be
exposed in the `/config` endpoint.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md
Comment (stevesg) [inline CHANGELOG.md:26]: Don't forget the PR number (yes, it is a chicken/egg situation)

```suggestion
* [ENHANCEMENT] Made handler function for `/config` endpoint overrideable so it can be extended by downstream projects. #4225
```

File: pkg/api/api.go
Comment (stevesg) [inline pkg/api/api.go:44]: I'm curious why this isn't the same as function signature as `DefaultConfigHandler`? I would assume if a client wanted to implement a custom version, they might want to replicate the behavior of the old `configHandler` (and customize).

File: pkg/api/handlers.go
Comment (stevesg) [inline pkg/api/handlers.go:114]: I'm curious (again) why this needs to be lazy/late evaluated (Not a leading question, genuinely curious) - might be nice not to have to do this unless really necessary, as there becomes a burden to support dynamically swapping out the provider.

I can see that downstream implementations would have to be careful to not e.g. start serving requests with one implementation, and then change the implementation, which would lead to confusion for the user having different results.

**Review by stevesg — COMMENTED** [review]
Looks sensible - just some minor comments and curiosities.

File: pkg/api/api.go
Comment (chaudum) [reply pkg/api/api.go:44]: You're totally right, it would be better to have one and the same interface.

File: pkg/api/handlers.go
Comment (chaudum) [reply pkg/api/handlers.go:114]: > I'm curious (again) why this needs to be lazy/late evaluated (Not a leading question, genuinely curious) - might be nice not to have to do this unless really necessary, as there becomes a burden to support dynamically swapping out the provider.

Right, I think it's not necessary to evaluate the handler lazily, as long as the module in the downstream project that override the handler has a dependency on the API module.

> I can see that downstream implementations would have to be careful to not e.g. start serving requests with one implementation, and then change the implementation, which would lead to confusion for the user having different results.

That's definitely not what one want! :D

**Comment by chaudum** [comment]
Hi @stevesg , would you mind having another look at the proposed changes? Thanks!

**Review by stevesg — APPROVED** [review]
All looks good to me, ncie work! Thanks for making the tweaks. Feel free to mark the conversations as resolved once you've done them.

Just FYI the maintainers also need to approve.

**Review by pstibrany — APPROVED** [review]
LGTM, thanks.

(I'm not sure this is "ENHANCEMENT" worth adding to CHANGELOG, as Cortex users don't see get anything out of this change.)

**Comment by chaudum** [comment]
> LGTM, thanks.
> 
> (I'm not sure this is "ENHANCEMENT" worth adding to CHANGELOG, as Cortex users don't see get anything out of this change.)

Thanks for the approval! :) I'm happy to remove it if you think it's not worth the mention.

**Review by pracucci — APPROVED** [review]
LGTM (modulo a nit). Thanks!

**Pull request merged**
